Chat refactor + persistent backing chat users

gabek opened the pull request:
This PR includes a complete rewrite of the chat server with some frontend changes to accommodate it.  It's a lot, so I'd appreciate any help testing it.  Luckily it's pretty easy to test and doesn't take much time to play around with the changes.  See the testing steps below.

# Overview of major changes

- Chat server is completely rewritten.
- Turning off chat in the settings now completely removes the chat server functionality, not just hides it from the UI.
- You must now be a registered user to chat or get chat history.
- Registration takes place in the background and is not an interactive thing.
- Users can be banned from chat.
- Number of concurrent connections to chat is greatly increased.
- Old chat history is deleted so we don't keep more data than needed.
- See "Detailed notes" on the bottom of this PR for more info.

# Things to keep in mind

- When you are blocked from chat, while technically you can get unblocked, a boolean is set in your localstorage that will continue to stop you from reconnecting in the frontend. I didn't want to add more complexity and APIs around a chat client having to check if its been unblocked.
- You could wipe your localstorage, open a different browser, or open an incognito window to get around being banned.  We do not currently store IP addresses.  Originally I was going to, but I really didn't want to store that identifying information.  If in the future we need to add this additional layer of blocking we can, but users could always use VPNs or Tor to get around that, too.
- You cannot access `/api/chat` without an access token, just to make it a little bit more inconvenient if somebody wanted to harvest chat history.  [3rd parties have hopefully seen the notice around this](https://owncast.online/news/20210615-apis-future-releases/).

# Testing

The following sections walks you through some specific items to test.

## To begin

1. Since this test includes blocking you from chat you might want to test in an incognito window so you don't have to deal with blocked state later.
1. Run from the **`develop` branch**.
1. Add a test access token in the admin from the **`develop` branch**.
1. **Change to this branch** and run the new branch with the bundled admin: `go run main.go pkged.go`.
1. Make sure there are no errors.
1. Start a stream.

## Chat

1. Verify you successfully connect to chat.
1. Send some chat messages.
1. Make sure your username is the same as it was in 0.0.7.
1. Verify you can change your name to something.
1. Try changing your name to `owncast` and verify it tells you that is disallowed.
1. Reload the page and make sure you can still chat and your user is the same.

## Chat admin

1. Verify you show up as a connected user under the chat -> users page.
1. Block your chat user from this page.
1. Verify you get disconnected from chat and it gets hidden.
1. Refresh the page. Verify chat is still not accessible.
1. Open the page in a new incognito window. Verify the chat messages from your blocked user are hidden.

## Access tokens

- In the admin make sure the test access token you created above still shows up. 

----
# Detailed notes

- A user access token is required to connect to the websocket or to hit the chat history API.
- When a user first visits the Owncast page they register for an access token and it's saved to browser local storage.
- If they already had a username set from a previous visit to Owncast that name will be passed when registering and that name will continue to be used.
- Default usernames are no longer `userxx`, but instead a server-generated name, Docker style, like `excellent-pizza`.
- External/3rd party access tokens have been converted to be treated the same as any other chat user, except they're meant to be used pragmatically.  So you can think of them as "API users" .
- 3rd party integrations can take advantage of a new API where they can send chat messages as themselves. This makes sense for bots.
- The chat server now increases its max connection count to 70% of the process max connection count.  Most people were limited at 1024 concurrent connections before and now they're not.  Max still depends on how their operating system is configured.
- User colors are persistent between user name changes.
- You can specify a list of names that are prohibited from being used.  Defaults are: "owncast", "operator", "admin", "system"
- User's name history and user creation time is tracked to help validity of users.
- You can block users from accessing the chat via the admin.
- When a user is blocked all of their messages are marked as hidden and they are forcefully disconnected, and chat removed from their UI.
- Old chat messages are completely deleted from the database in order to reduce the amount of history we have for privacy and data exposure reasons.
- Added a simple chat load test that will connect 5000 users and have them send messages. This replaces the old artillery-based one since that was too limited.
- Websocket has been moved to a more obvious `/ws` path.

gabek committed:
First pass at chat user registration and validation

gabek committed:
Disable chat if the user is disabled/blocked or the server hits max connections

gabek committed:
Handle dropping sockets if chat is disabled

gabek committed:
Fix origin in automated chat test

gabek committed:
Work for updated chat moderation

gabek committed:
Chat message markdown rendering and fix tests

gabek committed:
Put /api/chat behind a chat user access token. Closes #1085

gabek committed:
Reject blocked username changes

gabek committed:
More WIP moderation

gabek committed:
Defer configuring chat until we know if it is enabled. Closes #1135

gabek committed:
chat user blocking. Closes #1096

gabek committed:
Add tests around user access for #1096

gabek committed:
Add external integration chat message API + update integration auth middleware to pass along integration name. Closes #1092

gabek committed:
Delete old chat messages from db as to not hold on to excessive data. Closes #1152

gabek committed:
Add schema migration for messages. Closes #1155

Commit:
Commit updated API documentation

gabek committed:
Add chat load test

gabek committed:
Shared db mutex and db optimizations

gabek committed:
Simplify past display name handling

gabek committed:
Use a new test db for each test run

gabek committed:
Wire up the external messages actions + add tests for them

gabek committed:
Move access tokens to be actual users

gabek committed:
Run message pruning at launch + fix comparison

gabek committed:
Do not return API users in disabled users response

gabek committed:
Fix incorrect highlighting. Closes #1160

gabek committed:
Consolidate user table statements

gabek committed:
Set the max process connection limit to 70% of maximum

gabek committed:
Fix wrong old display name being returned in name change event

gabek committed:
Delete the old chat server files

gabek committed:
Wire back up the webhooks

gabek committed:
Remove unused

gabek committed:
Invalidate user cache on changes

gabek committed:
Do not send rendered body as RawBody

gabek committed:
Some cleanup

gabek committed:
Standardize names for external API users to ExternalAPIUser

gabek committed:
Do not log token

gabek committed:
Checkout branch when building admin for testing

gabek committed:
Bundle in dev admin for testing

gabek committed:
Some cleanup

gabek committed:
Cleanup js logs

gabek committed:
Cleanup and standardize event names

gabek committed:
Clean up some logging

gabek committed:
Update API spec. Closes #1133

Commit:
Commit updated API documentation

gabek committed:
Change paths to be better named

Commit:
Commit updated API documentation

gabek committed:
Update admin bundle

gabek committed:
Fix duplicate event name

gabek committed:
Rename scope var

gabek committed:
Update admin bundle

gingervitis commented:
Going to this page `/admin/access-tokens/` right after switching over to this branch gives an error.
<img width="1425" alt="image" src="https://user-images.githubusercontent.com/9563255/125173039-bfc75d80-e171-11eb-9832-e06f69b28917.png">

gingervitis commented:
Changing your name to `"owncast"` - new name does not go away after error message is returned.

<img width="138" alt="image" src="https://user-images.githubusercontent.com/9563255/125173163-bab6de00-e172-11eb-918f-6c3c513d3a54.png">

The forbidden "owncast" name remains even after reloading the page.

However, chatting will display your old user name.
<img width="356" alt="image" src="https://user-images.githubusercontent.com/9563255/125173228-16816700-e173-11eb-9631-1257a711f2cf.png">

gingervitis commented:
Some funkiness:

I have 2 different browsers going. "User 301" and "User 11".

1. From the Admin I banned "User 301".
2. Go to User 301's page. 301's chat clears out. But the panel is still there, and can still input and send messages. But this JS error does show up for it, post-ban.
<img width="697" alt="image" src="https://user-images.githubusercontent.com/9563255/125173703-63fed380-e175-11eb-90b9-c9c6e34b8a72.png">

3. Go to User 11's page. 11's chat panel has disappeared. (and it looks not-great)
4. Go reload the Admin chat user's page. 301 shows in both the Users table and the Blocked table. User 11 has disappeared.
<img width="542" alt="image" src="https://user-images.githubusercontent.com/9563255/125173552-d15e3480-e174-11eb-949e-316c2bc75a5c.png">


Other funkiness: 
I noticed that in the Admin users table, the ordering of User 301 and User 11 would keep switching in the table. There might be something going on there, which may have confused which user actually got banned.

gingervitis commented:
This happens when trying to auto complete names (does not happen in `develop`)
<img width="357" alt="image" src="https://user-images.githubusercontent.com/9563255/125174297-1e440a00-e179-11eb-922a-7e6bbd1d4e3d.png">

gingervitis commented:
I restarted a new test with 2 browsers again , and did not come across these errors the 2nd time.

https://github.com/owncast/owncast/pull/1163#issuecomment-877687381

jeyemwey commented on build/admin/bundleAdmin.sh:
```diff
@@ -18,6 +18,7 @@ trap shutdown INT TERM ABRT EXIT
 echo "Cloning owncast admin into $INSTALL_TEMP_DIRECTORY..."
 git clone https://github.com/owncast/owncast-admin 2> /dev/null
 cd owncast-admin
+git checkout gek/chat-user-refactor
```
Remove when merging into `develop`?

gabek replied:
Yup!

jeyemwey commented on controllers/admin/serverConfig.go:
```diff
@@ -15,6 +16,7 @@ import (
 // GetServerConfig gets the config details of the server.
 func GetServerConfig(w http.ResponseWriter, r *http.Request) {
 	ffmpeg := utils.ValidatedFfmpegPath(data.GetFfMpegPath())
+	usernameBlocklist := strings.Join(data.GetUsernameBlocklist(), ",")
```
Why is `serverConfigAdminResponse.UsernameBlocklist` not a `[]string` but just a string? JSON offers a list encoding, and it usually makes sense to use it (if only for trimming spaces).

gabek replied:
It's being stored as a string, so one side or the other needs to do the translation, so I just took the route of the least amount of code and allowed it to be sent as a string.  I went ahead and changed it to be an array over the wire, but the end result is the same, with the trimming of whitespace still happening [during the comparison](https://github.com/owncast/owncast/pull/1163/files#diff-249ac4ddbe2945013dfb90c7fc32f299ea8e9a3326e67033eb14b476323ff004R25-R28).

jeyemwey commented on core/chat/events.go:
```diff
@@ -0,0 +1,98 @@
+	savedUser.DisplayName = receivedEvent.NewName
+
+	// Update the connected clients associated user with the new name
+	eventData.client.User = savedUser
```
Swap, so the comments stay correct:

```suggestion
	// Update the connected clients associated user with the new name
	eventData.client.User = savedUser

	// Send chat event letting everyone about about the name change
	savedUser.DisplayName = receivedEvent.NewName
```

jeyemwey commented on core/chat/events/events.go:
```diff
@@ -12,51 +12,72 @@ import (
+
+// Event is any kind of event.  A type is required to be specified.
+type Event struct {
+	Type      string    `json:"type"`
```
Can we pin this to be of type `EventType`?

```suggestion
	Type      EventType `json:"type"`
```

gabek committed:
Move connected clients controller into admin package

jeyemwey commented on core/chat/events/nameChangeEvent.go:
```diff
@@ -0,0 +1,26 @@
+// GetBroadcastPayload will return the object to send to all chat users.
+func (e *NameChangeBroadcast) GetBroadcastPayload() EventPayload {
+	return EventPayload{
+		"id":        e.Id,
+		"timestamp": e.Timestamp,
+		"user":      e.User,
```
Looking at the other events, this looks like a lot of duplicated code. Can we create a method which creates an `EventPayload` from any `e Event`, and add items to the `map[string]interface{}` later on?

gabek replied:
I did this on purpose so the outbound payloads were explicit and obvious and are not bound by an abstraction, since these values are specifically for client consumption.  I like being able to look at a specific event and know exactly how it's being generated, and it makes it clear when adding more in the future.  If we wanted to type the outbound events (like we do with the internal objects) then we can abstract it more, but I chose this flexibility over having an additional specific `BroadcastNameChangeEvent` type, even though that would also be fine.

jeyemwey commented on core/chat/events/systemMessageEvent.go:
```diff
@@ -0,0 +1,26 @@
+// SystemMessageEvent will return the object to send to all chat users.
+func (e *SystemMessageEvent) GetBroadcastPayload() EventPayload {
+	return EventPayload{
+		"id":        e.Id,
+		"timestamp": e.Timestamp,
+		"body":      e.Body,
```
See above.

jeyemwey commented on core/chat/events/userDisabledEvent.go:
```diff
@@ -0,0 +1,17 @@
+func (e *UserDisabledEvent) GetBroadcastPayload() EventPayload {
+	return EventPayload{
+		"type":      ErrorUserDisabled,
+		"id":        e.Id,
+		"timestamp": e.Timestamp,
+		"user":      e.User,
```
See above.

jeyemwey commented on core/chat/events/userJoinedEvent.go:
```diff
@@ -0,0 +1,17 @@
+func (e *UserJoinedEvent) GetBroadcastPayload() EventPayload {
+	return EventPayload{
+		"type":      UserJoined,
+		"id":        e.Id,
+		"timestamp": e.Timestamp,
+		"user":      e.User,
```
See above.

jeyemwey commented on core/chat/events/userMessageEvent.go:
```diff
@@ -0,0 +1,24 @@
+// GetBroadcastPayload will return the object to send to all chat users.
+func (e *UserMessageEvent) GetBroadcastPayload() EventPayload {
+	return EventPayload{
+		"id":        e.Id,
+		"timestamp": e.Timestamp,
+		"body":      e.Body,
+		"user":      e.User,
```
See above (except for the body).

gabek committed:
Fix collecting usernames for autocomplete purposes

gabek committed:
No longer generate username when it is empty

gabek committed:
Sort clients and users by timestamp

gabek committed:
Move file to admin controller package

jeyemwey commented on models/webhook.go:
```diff
@@ -25,7 +29,8 @@ var validEvents = []EventType{
 	for _, event := range events {
-		if !findItemInSlice(validEvents, event) {
+		_, foundInSlice := utils.FindInSlice(validEvents, event)
+		if !foundInSlice {
```
```suggestion
		if _, foundInSlice := utils.FindInSlice(validEvents, event); !foundInSlice {
```

jeyemwey commented on test/automated/chat.test.js:
```diff
@@ -2,52 +2,43 @@ const { test } = require('@jest/globals');
-const messageMarkdown = '<p>' + message + ' <em>and some markdown too</em></p>'
-const date = new Date().toISOString();
+// const messageRaw = message + ' *and some markdown too*';
+// const messageMarkdown = '<p>' + message + ' <em>and some markdown too</em></p>'
```
Remove code dumps

```suggestion
```

jeyemwey commented on test/automated/chat.test.js:
```diff
@@ -2,52 +2,43 @@ const { test } = require('@jest/globals');
+        return (msg.body = testMessage.body);
+      })[0];
+
+      expect(message.body).toBe(message.body);
```
Will always be true. You probably want to diff against the `testMessage.body`?

jeyemwey commented on test/automated/chat.test.js:
```diff
@@ -2,52 +2,43 @@ const { test } = require('@jest/globals');
+    .expect(200)
+    .then((res) => {
+      const message = res.body.filter(function (msg) {
+        return (msg.body = testMessage.body);
```
Missing equal sign, thus not filtering properly.

gabek replied:
Great find, fixed!

jeyemwey commented on test/automated/chat.test.js:
```diff
@@ -2,52 +2,43 @@ const { test } = require('@jest/globals');
+
+      expect(message.body).toBe(message.body);
+      expect(message.user.displayName).toBe(userDisplayName);
+      // expect(message.date).toBe(testMessage.date);
```
Remove commented code.
```suggestion
```

jeyemwey commented on test/automated/chat.test.js:
```diff
@@ -2,52 +2,43 @@ const { test } = require('@jest/globals');
-            done();
-        });
-    }
+test('can send a chat message', async (done) => {
```
The `can send a chat message` and `can fetch a chat message` test need to work together to work properly (which you can smell from the global `var userDisplayName`. IMO, it would be better to wrap both tests in a `describe` block, so it's obvious that they work with the same data.

gabek replied:
As far as I've seen tests have always run in order, so I've never seen any issues from it.  But I'm no expert in Jest.  Just for clarification I checked with @gingervitis and she believes they always run in order.  This could be updated in the future if we want to spend some time grouping them together.  But all the tests currently work this way, and they're not broken now and are performing as expected.

jeyemwey commented on test/automated/chatmoderation.test.js:
```diff
@@ -2,38 +2,30 @@ const { test } = require('@jest/globals');
+test('can send a chat message', async (done) => {
+    const registration = await registerChat();
+    const accessToken = registration.accessToken;
+    userDisplayName = registration.displayName;
```
`userDisplayName` is not declared for this test, is never used later. I'd vote to remove it from the test.

jeyemwey commented on test/automated/chatusers.test.js:
```diff
@@ -0,0 +1,73 @@
+    type: 'CHAT',
+};
+
+var userId
+var accessToken
+test('can register a user', async (done) => {
+    const registration = await registerChat();
+    userId = registration.id;
+    accessToken = registration.accessToken;
+    done();
+});
```
Same as with `test/automated/chat.test.js`, the tests are not independent and need to run in a specific order. It would be more obvious what is happening where if we were using `describe` or `beforeEach`/`beforeAll`.

jeyemwey commented on test/load/chatLoadTest.js:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+async function sendMessage() {
```
From a review POV, it's not clear what is happening when. Is this function only running once? Is it calling itself (with a timeout)? What is the `send` function doing inside of it?

gabek replied:
I renamed and added comments it to be clearer.  I could totally spend a bunch of time building a far cleaner script, but this just needs to do one specific thing to assist in testing, so it was thrown together for that because the load testing framework wouldn't allow me to do what I needed to do.  Hopefully we can get rid of it if Artillery's update comes out some day.

jeyemwey commented on utils/phraseGenerator.go:
```diff
@@ -0,0 +1,636 @@
+package utils
```
This is the Docker list, isn't it? Is there a reason why we can't import their package, but have to copy it over? Is that fine with Docker's (and our) IP Licensing?

gabek replied:
I wanted our own list we could customize.  No headers were stripped from the original.

jeyemwey reviewed (commented):
Hi Gabe, I've added a bunch of comments regarding the Go code and some testing bits. Mostly smaller things, so no worries.

I did not touch the `webroot/` folder, and most of the business logic, so it would be helpful if someone would look on that with a clear mind.

@ whoever shies away from the line count: The PR also includes 1,200 diff lines of Documentation, 6,300 diffs of a `package-lock.json` and 600 lines for a phrase generator. So it's "only" around 2,200 additions that need to be looked at :joy:

gabek committed:
Swap, so the comments stay correct

Co-authored-by: Jannik <jannik@outlook.com>

gabek committed:
Use explicit type alias

Co-authored-by: Jannik <jannik@outlook.com>

gabek committed:
Remove commented code.

Co-authored-by: Jannik <jannik@outlook.com>

gabek committed:
Cleanup test

gabek committed:
Remove some extra logging

gabek committed:
Add some clarity

gabek committed:
Update dev instance of admin for testing

gingervitis commented:
`previousNames` for a user keeps coming back `null`

gabek committed:
Until needed do not return IP address with client list

gabek committed:
Consolidate lines

Co-authored-by: Jannik <jannik@outlook.com>

gabek committed:
Remove commented unused vars

Co-authored-by: Jannik <jannik@outlook.com>

gabek committed:
Fix typo of wrong var

gabek committed:
Typo led to a bad test. Fix typo and fix test.

gabek committed:
Guard against the socket reconnecting on error if previously set to shutdown

gabek committed:
Do not log access tokens

gabek committed:
Return success message on enable/disable user

gabek committed:
Clean up some inactionable error messages. Sent ban message. Sort banned users.

gabek committed:
Unused

gabek committed:
guard against nil clients

gingervitis committed:
fix styling for when chat is completely disabled

gabek committed:
Update dev admin bundle

gabek committed:
Fix broken workflow due to typo

gabek committed:
Troubleshoot workflow

gabek committed:
Do not unhide messages when unblocking user just to be safe. Send removal action from the controller

gabek committed:
Add convinience function for getting active connections for a single user

gabek committed:
Lock db on these mutations

gabek committed:
Cleanup force disconnect using GetClientsForUser and capture client reference explicitly

gabek committed:
No longer re-showing banned user messages for safety. Removing this test.

gabek committed:
Remove no longer needed comment

gabek committed:
Tweaks to forbidden username handling.

- Standardize naming to not use "block" but "forbidden" instead.
- Pass array over the wire instead of string.
- Add API test
- Fix default list incorrectly being appended to custom list.

gabek committed:
Logging cleanup

gabek committed:
Update dev admin bundle

gabek committed:
Add an artificial delay in order to visually see message being hidden when testing

gabek committed:
Remove the user cache as it is a premature optimization

gabek committed:
When connected to chat let the user know their current user details to sync the username in the UI

gabek committed:
On connected send current display name back to client.
- Move name change out of chat component.
- Add additional event type constants.

dependabot[bot] committed:
Bump htm from 3.0.4 to 3.1.0 in /build/javascript (#1181)

* Bump htm from 3.0.4 to 3.1.0 in /build/javascript

Bumps [htm](https://github.com/developit/htm) from 3.0.4 to 3.1.0.
- [Release notes](https://github.com/developit/htm/releases)
- [Commits](https://github.com/developit/htm/compare/3.0.4...3.1.0)

---
updated-dependencies:
- dependency-name: htm
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Run npm run build and update libraries

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Gabe Kangas <gabek@real-ity.com>

Commit:
Commit updated Javascript packages

gabek committed:
Re-send current user info when a rejected name change takes place

gabek committed:
All socket writes should be through the send chan and not directly

gabek committed:
Seed the random generator

gabek committed:
Add keys and indexes to users table

gingervitis committed:
a util to generate consistent emoji markup

gingervitis committed:
console clean up

gabek committed:
mod tidy

Commit:
Commit updated API documentation

gabek committed:
Handle the max payload size of a socket message.
- Only close socket if x2 greater than the max size.
- Send the user a message if a message is too large.
- Surface the max size in bytes in the config.

gabek committed:
Update admin bundle

gabek committed:
Force all events to be sent in their own socket message and do not concatinate in a single message

gabek committed:
Update chat embed to register for access token

gabek committed:
Use a different access token for embed chat

gabek committed:
Update the chat message bubble background color to be bolder

gingervitis committed:
add base tag to open links in new window, closes #1220

gingervitis committed:
Merge branch 'gw/chat/add-blank-tolinks' into gek/chat-rebuild-with-access-tokens

MFTabriz committed:
Support text input of :emoji: in chat (#1190)

* Initial implementation of emoji injection

* fix bookkeeping with multiple emoji

* make the emoji lookup case-insensitive

* try another solution for Caretposition

* add title to emojis

minor refactoring

* bind moji injection to InputKeyUp

* simplify the code

replace all found emojis

* inject emoji if the modifer is released earlier

* more efficient emoji tag search

* use json emoji.emoji as url

* use createEmojiMarkup()

* move emojify() to chat.js

* emojify on paste

* cleanup emoji titles in paste

* update inputText in InputKeyup

* mark emoji titles with 2*zwnj

this way paste cleanup will not interfere with text which include zwnj

* emoji should not change the inputText

gabek committed:
Do not show join messages when chat is offline. Closes #1224
- Show stream starting/ending messages in chat.
- When stream starts show everyone the welcome message.

gabek committed:
Force scrolling chat to bottom after history is populated regardless of scroll position. Closes https://github.com/owncast/owncast/issues/1222

gingervitis committed:
use maxSocketPayloadSize to calculate total bytes of message payload (#1221)

* utilize maxSocketPayloadSize from config; update chatInput to calculate based on that value instead of text value; remove usage of inputText for counting

* add a buffer to account for entire websocket payload for message char counting; trim nbsp;'s from ends of messages when calculating count

Co-authored-by: Gabe Kangas <gabek@real-ity.com>
